pkg/driver/driver_packet.go

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright (c) 2017 SAP SE or an SAP affiliate company. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package driver contains the cloud provider specific implementations to manage machines
+package driver
+
+import (
+	"fmt"
+	"strings"
+
+	v1alpha1 "github.com/gardener/machine-controller-manager/pkg/apis/machine/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/golang/glog"
+	"github.com/packethost/packngo"
+)
+
+// PacketDriver is the driver struct for holding Packet machine information
+type PacketDriver struct {
+	PacketMachineClass *v1alpha1.PacketMachineClass
+	CloudConfig        *corev1.Secret
+	UserData           string
+	MachineID          string
+	MachineName        string
+}
+
+// NewPacketDriver returns an empty PacketDriver object
+func NewPacketDriver(create func() (string, error), delete func() error, existing func() (string, error)) Driver {
+	return &PacketDriver{}
+}
+
+// Create method is used to create a Packet machine
+func (d *PacketDriver) Create() (string, string, error) {
+
+	svc := d.createSVC()
+	// packet tags are strings only
+	tags := packetTagsMapToString(d.PacketMachineClass.Spec.Tags)
+	createRequest := &packngo.DeviceCreateRequest{
+		UserData:  d.UserData,
+		Plan:      d.PacketMachineClass.Spec.MachineType,
+		ProjectID: d.PacketMachineClass.Spec.ProjectID,
+		Facility:  d.PacketMachineClass.Spec.Facility,
+		OS:        d.PacketMachineClass.Spec.OS,
+		Tags:      tags,
+	}
+
+	device, _, err := svc.Devices.Create(createRequest)
+	if err != nil {
+		glog.Errorf("Could not create machine: %v", err)
+		return "", "", err
+	}
+	return d.encodeMachineID(device.Facility.ID, device.ID), device.Hostname, nil
+}
+
+// Delete method is used to delete a Packet machine
+func (d *PacketDriver) Delete() error {
+
```

Comment:
Does it make sense to check if the VM actually exists, before trying to delete it? Refer [this](https://github.com/gardener/machine-controller-manager/blob/master/pkg/driver/driver_aws.go#L138-L145). Even checking based on SDK call return would suffice.

Reply:
OK checking.

Reply:
My understanding then is that `Delete()` should be idempotent? If I call `Delete()` on a device that no longer exists (or never did), it should return `nil` rather than an error? When I call delete:

* if after call, no such device will exist (whether because it didn't exist before, or because it successfully deleted): return `nil`
* if delete fails, return `error`

Is that right?

Reply:
In that case, I am going to do it slightly differently. The packet go client just wraps a REST interface. I am going to leave it as is, but capture `404` returns and handle them.

Reply:
Thanks 👍  Capturing a 404 should also work.